### PR TITLE
[FLINK-20888][runtime] Close outputs in OperatorChain.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

Currently, `ContinuousFileReaderOperator` closes the following chained operator on reaching `endOfInput`.

## Brief change log

- Make sure that everything is closed where it's owned - not more or less.

## Verifying this change

Added ContinuousFileReaderOperatorITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by AHeise** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java [hunk 1]
```diff
@@ -124,7 +124,7 @@ public AbstractStreamOperatorV2(StreamOperatorParameters<OUT> parameters, int nu
             operatorMetricGroup = null;
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java:124]:
```
operatorMetricGroup = null;
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ChainingOutput.java [hunk 2]
```diff
@@ -129,13 +126,7 @@ public void emitLatencyMarker(LatencyMarker latencyMarker) {
 
     @Override
     public void close() {
-        try {
-            if (closeable != null) {
-                closeable.close();
-            }
-        } catch (Exception e) {
-            throw new ExceptionInChainedOperatorException(e);
-        }
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ChainingOutput.java:138]: unit test?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 3]
```diff
@@ -335,7 +335,11 @@ public void open() throws Exception {}
      * @throws Exception An exception in this method causes the operator to fail.
      */
     @Override
-    public void close() throws Exception {}
+    public void close() throws Exception {
+        if (output != null) {
+            output.close();
+        }
+    }
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:342]: I think the `OperatorChain` should be the owner of the outputs. He is referencing some of the created outputs and using them for other things:
- registering some metrics (maybe can be safely ignored?)
- `OperatorChain#chainedSources` are also referencing some outputs

Hence I would suggest that `OperatorChain#releaseOutputs()` should be the guy to close all outputs that were created inside that `OperatorChain`.

Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: Yes, I agree.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ChainingOutput.java (hunk 2 above)
Comment (AHeise) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ChainingOutput.java:138]: I don't think, I need to test an empty method. I have added a comment that only things that are owned should be closed with a reference to this ticket to protect against regressions.

**Review by pnowojski — APPROVED** [review]
lgtm!

**Pull request merged**